From 472da265801c600d02dd2af141c79a86b6f50ca0 Mon Sep 17 00:00:00 2001
From: yangwei1 <yangwei1@eswincomputing.com>
Date: Tue, 7 Jan 2025 14:16:06 +0800
Subject: [PATCH 375/416] WIN2030-15186:feat:support tps549d22 amd tmp102 in
 7702T

Changelogs:
	cp dts form eic7702-evb-a1.dts to eic7702-tb.dts;support tps549d22 to supply for npu;supprot tmp102 to get environment temperature

Change-Id: I2a91636252643f731ad09e986f2de2bd149d2428
Signed-off-by: yangwei1 <yangwei1@eswincomputing.com>
---
 arch/riscv/boot/dts/eswin/eic7702-tb.dts | 1620 ++++++++++++++++++++++
 arch/riscv/configs/eic7700_defconfig     |    4 +-
 arch/riscv/configs/eic7702_defconfig     |    1 +
 drivers/regulator/Kconfig                |    7 +-
 drivers/regulator/Makefile               |    1 +
 drivers/regulator/tps549d22-regulator.c  |  580 ++++++++
 6 files changed, 2211 insertions(+), 2 deletions(-)
 create mode 100644 arch/riscv/boot/dts/eswin/eic7702-tb.dts
 create mode 100644 drivers/regulator/tps549d22-regulator.c

diff --git a/arch/riscv/boot/dts/eswin/eic7702-tb.dts b/arch/riscv/boot/dts/eswin/eic7702-tb.dts
new file mode 100644
index 000000000000..2355bd3f422d
--- /dev/null
+++ b/arch/riscv/boot/dts/eswin/eic7702-tb.dts
@@ -0,0 +1,1620 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Device Tree file for Eswin EIC7700 SoC.
+ *
+ * Copyright 2024, Beijing ESWIN Computing Technology Co., Ltd.. All rights reserved.
+ * SPDX-License-Identifier: GPL-2.0
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/dts-v1/;
+
+#define RTCCLK_FREQ		1000000
+#define LSPCLK_FREQ		200000000
+
+/* If wanna enable ECC capability of DDR, should reserve highest zone of 1/8 all space for it */
+#define MEMORY_SIZE_H		0x4
+#define MEMORY_SIZE_L		0x0
+#define CMA_SIZE		0x20000000
+
+#include "eswin-win2030-die0-soc.dtsi"
+#include "eswin-win2030-die1-soc.dtsi"
+
+#include "eic7x-die0-pinctrl.dtsi"
+#include "eic7x-die1-pinctrl.dtsi"
+#include <dt-bindings/interrupt-controller/irq.h>
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/* Clock frequency (in Hz) of the PCB crystal for rtcclk */
+
+/ {
+	#address-cells = <2>;
+	#size-cells = <2>;
+	model = "ESWIN EIC7702 TB";
+	compatible = "sifive,hifive-unmatched-a00", "sifive,fu740-c000",
+		     "sifive,fu740", "eswin,eic7700";
+
+	aliases {
+		serial0 = &d0_uart0;
+		ethernet0 = &d0_gmac0;
+		ethernet1 = &d0_gmac1;
+		ethernet2 = &d1_gmac0;
+		ethernet3 = &d1_gmac1;
+		rtc0 = &die0_rtc;
+		rtc1 = &d1_rtc;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	cpus {
+		timebase-frequency = <RTCCLK_FREQ>;
+	};
+
+	memory@59000000 {
+		device_type = "memory";
+		reg = <0x0 0x59000000 0x0 0x400000>;
+		numa-node-id = <0>;
+	};
+
+	memory@79000000 {
+		device_type = "memory";
+		reg = <0x0 0x79000000 0x0 0x400000>;
+		numa-node-id = <1>;
+	};
+
+	memory@80000000 {
+		device_type = "memory";
+		reg = <0x0 0x80000000 MEMORY_SIZE_H MEMORY_SIZE_L>;
+		numa-node-id = <0>;
+	};
+
+	memory@2000000000 {
+		device_type = "memory";
+		reg = <0x20 0x00000000 MEMORY_SIZE_H MEMORY_SIZE_L>;
+		numa-node-id = <1>;
+	};
+
+	reserved-memory {
+		#address-cells = <2>;
+		#size-cells = <2>;
+		ranges;
+
+		linux,cma {
+			compatible = "shared-dma-pool";
+			reusable;
+			size = <0x0 CMA_SIZE>;
+			alignment = <0x0 0x1000>;
+			alloc-ranges = <0x20 0x00000000 MEMORY_SIZE_H MEMORY_SIZE_L>;
+			linux,cma-default;
+		};
+
+		npu0_reserved: sprammemory@59000000 {
+			no-map;
+			reg = <0x0 0x59000000 0x0 0x400000>;
+		};
+
+		npu1_reserved: sprammemory@79000000 {
+			no-map;
+			reg = <0x0 0x79000000 0x0 0x400000>;
+		};
+
+		g2d_4GB_boundary_reserved_4k {
+			no-map;
+			reg = <0x0 0xfffff000 0x0 0x1000>;
+		};
+
+		g2d_8GB_boundary_reserved_4k {
+			no-map;
+			reg = <0x1 0xfffff000 0x0 0x1000>;
+		};
+
+		g2d_12GB_boundary_reserved_4k {
+			no-map;
+			reg = <0x2 0xfffff000 0x0 0x1000>;
+		};
+
+		mmz_nid_0_part_0 {
+			compatible = "eswin-reserve-memory";
+			reg = <0x3 0x0 0x1 0x80000000>;
+			no-map;
+		};
+
+		nid_1_zero_device_simu: nid_1_zero_device_simu@2040000000 {
+			no-map;
+			reg = <0x20 0x40000000 0x0 0x400000>;
+		};
+
+		d1_g2d_4GB_boundary_reserved_4k {
+			no-map;
+			reg = <0x20 0xfffff000 0x0 0x1000>;
+		};
+
+		d1_g2d_8GB_boundary_reserved_4k {
+			no-map;
+			reg = <0x21 0xfffff000 0x0 0x1000>;
+		};
+
+		mmz_nid_1_part_0 {
+			compatible = "eswin-reserve-memory";
+			reg = <0x22 0x80000000 0x1 0x80000000>;
+			no-map;
+		};
+	};
+
+	distance-map {
+		compatible = "numa-distance-map-v1";
+		distance-matrix = <0 0 10>,
+				<0 1 100>,
+				<1 0 100>,
+				<1 1 10>;
+	};
+
+	/* die1 gpio68 system led */
+	leds {
+		compatible = "gpio-leds";
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_die1_gpio68_default>;
+
+		gpio-68 {
+			gpios = <&d1_portc 4 GPIO_ACTIVE_LOW>;
+			label = "heartbeat";
+			linux,default-trigger = "heartbeat";
+		};
+	};
+
+};
+
+&D1CACHE {
+	status = "okay";
+	d1_zero_device: zero-device@3a000000 {
+		compatible = "l3,zero-device";
+		/*reg = <0x0 0x3a000000 0x0 0x400000>;*/
+		memory-region = <&nid_1_zero_device_simu>;
+	};
+};
+
+&d0_clock {
+	status = "okay";
+};
+&d1_clock {
+	status = "okay";
+};
+
+&d0_reset {
+	status = "okay";
+};
+&d1_reset {
+	status = "okay";
+};
+
+&d0_pmu {
+	status = "okay";
+};
+&d1_pmu {
+	status = "disabled";
+};
+
+&ddr0 {
+	status = "okay";
+};
+
+&ddr1 {
+	status = "okay";
+};
+
+&d1_ddr0 {
+	status = "okay";
+};
+
+&d1_ddr1 {
+	status = "okay";
+};
+
+&smmu0 {
+	status = "okay";
+};
+&smmu1 {
+	status = "okay";
+};
+
+&smmu_pmu0 {
+	status = "disabled";
+};
+&smmu_pmu1 {
+	status = "disabled";
+};
+
+
+&dev_foo_a {
+	status = "okay";
+};
+
+&d0_cfg_noc {
+	status = "okay";
+};
+
+&d0_llc_noc {
+	status = "okay";
+	stat,0 = "TracePort:ddr0_p0_req";
+	stat,1 = "TracePort:ddr1_p0_req";
+	//latency,0 = "TracePort:llcnoc_trans_probe";
+	//pending,0 = "TracePort:llcnoc_trans_probe";
+};
+
+&d0_sys_noc {
+	status = "okay";
+
+	//eswin,DSPT-qos-owner;
+	//eswin,NPU-qos-owner;
+	//eswin,SPISLV_TBU3-qos-owner;
+
+	stat,0 = "TracePort:ddr0_p1_req",
+		 "InitFlow:mcput_snoc_mp/I/0";
+
+	stat,1 = "TracePort:ddr0_p2_req",
+		 "InitFlow:dspt_snoc/I/0",
+		 "AddrBase:0x81000000", "AddrSize:0x30",
+		 "Opcode:RdWrLockUrg", "Status:ReqRsp", "Length:0x8000", "Urgency:0x0";
+
+	stat,2 = "TracePort:ddr1_p1_req",
+		 "Status:Req", "AddrSize:0x28";
+
+	stat,3 = "TracePort:ddr1_p2_req";
+
+	latency,0 = "TracePort:sysnoc_trans_probe_0", "AddrSize:0x0";
+	latency,1 = "TracePort:sysnoc_trans_probe_1","Mode:latency","AddrBase:0x82000000","AddrSize:0x28","Opcode:RdWr";
+	//latency,2 = "TracePort:sysnoc_trans_probe_2";
+
+	//pending,0 = "TracePort:sysnoc_trans_probe_0";
+	//pending,1 = "TracePort:sysnoc_trans_probe_1","Mode:latency","AddrBase:0x82000000","AddrSize:0x0","Opcode:RdWr";
+	pending,0 = "TracePort:sysnoc_trans_probe_2", "AddrSize:0x3";
+};
+
+&d0_media_noc {
+	status = "okay";
+
+	//eswin,GPU-qos-owner;
+	//eswin,TBU2-qos-owner;
+	//eswin,VC-qos-owner;
+
+	stat,0 = "TracePort:ddr0_p3_req";
+	stat,1 = "TracePort:ddr1_p3_req";
+	//latency,0 = "TracePort:mnoc_trans_probe";
+	//pending,0 = "TracePort:mnoc_trans_probe";
+};
+
+&d0_realtime_noc {
+	status = "okay";
+
+	//eswin,TBU0-qos-owner;
+	//eswin,VO-qos-owner;
+
+	stat,0 = "TracePort:ddr0_p4_req";
+	stat,1 = "TracePort:ddr1_p4_req";
+	//latency,0 = "TracePort:rnoc_trans_probe";
+	//pending,0 = "TracePort:rnoc_trans_probe";
+};
+
+&d0_noc_wdt {
+	status = "okay";
+};
+
+&d1_cfg_noc {
+	status = "okay";
+};
+
+&d1_llc_noc {
+	status = "okay";
+	stat,0 = "TracePort:ddr0_p0_req";
+	stat,1 = "TracePort:ddr1_p0_req";
+	//latency,0 = "TracePort:llcnoc_trans_probe";
+	//pending,0 = "TracePort:llcnoc_trans_probe";
+};
+
+&d1_sys_noc {
+	status = "okay";
+
+	//eswin,DSPT-qos-owner;
+	//eswin,NPU-qos-owner;
+	//eswin,SPISLV_TBU3-qos-owner;
+
+	stat,0 = "TracePort:ddr0_p1_req",
+		 "InitFlow:mcput_snoc_mp/I/0";
+
+	stat,1 = "TracePort:ddr0_p2_req",
+		 "InitFlow:dspt_snoc/I/0",
+		 "AddrBase:0x81000000", "AddrSize:0x30",
+		 "Opcode:RdWrLockUrg", "Status:ReqRsp", "Length:0x8000", "Urgency:0x0";
+
+	stat,2 = "TracePort:ddr1_p1_req",
+		 "Status:Req", "AddrSize:0x28";
+
+	stat,3 = "TracePort:ddr1_p2_req";
+
+	latency,0 = "TracePort:sysnoc_trans_probe_0", "AddrSize:0x0";
+	latency,1 = "TracePort:sysnoc_trans_probe_1","Mode:latency","AddrBase:0x82000000","AddrSize:0x28","Opcode:RdWr";
+	//latency,2 = "TracePort:sysnoc_trans_probe_2";
+
+	//pending,0 = "TracePort:sysnoc_trans_probe_0";
+	//pending,1 = "TracePort:sysnoc_trans_probe_1","Mode:latency","AddrBase:0x82000000","AddrSize:0x0","Opcode:RdWr";
+	pending,0 = "TracePort:sysnoc_trans_probe_2", "AddrSize:0x3";
+};
+
+&d1_media_noc {
+	status = "okay";
+
+	//eswin,GPU-qos-owner;
+	//eswin,TBU2-qos-owner;
+	//eswin,VC-qos-owner;
+
+	stat,0 = "TracePort:ddr0_p3_req";
+	stat,1 = "TracePort:ddr1_p3_req";
+	//latency,0 = "TracePort:mnoc_trans_probe";
+	//pending,0 = "TracePort:mnoc_trans_probe";
+};
+
+&d1_realtime_noc {
+	status = "okay";
+
+	//eswin,TBU0-qos-owner;
+	//eswin,VO-qos-owner;
+
+	stat,0 = "TracePort:ddr0_p4_req";
+	stat,1 = "TracePort:ddr1_p4_req";
+	//latency,0 = "TracePort:rnoc_trans_probe";
+	//pending,0 = "TracePort:rnoc_trans_probe";
+};
+
+&d1_noc_wdt {
+	status = "okay";
+};
+
+&d0_ipc_scpu {
+	status = "okay";
+};
+&d1_ipc_scpu {
+	status = "okay";
+};
+
+&d0_lpcpu {
+	status = "okay";
+};
+&d1_lpcpu {
+	status = "disabled";
+};
+
+&pcie {
+	status = "okay";
+};
+&d1_pcie {
+	status = "okay";
+};
+
+&d0_npu{
+	status = "okay";
+};
+&d1_npu {
+	status = "okay";
+};
+
+&d0_dsp_subsys {
+	status = "okay";
+};
+&d1_dsp_subsys {
+	status = "okay";
+};
+
+&d0_dsp0 {
+	status = "okay";
+};
+&d1_dsp0 {
+	status = "okay";
+};
+
+&d0_dsp1 {
+	status = "okay";
+};
+&d1_dsp1 {
+	status = "okay";
+};
+
+&d0_dsp2 {
+	status = "okay";
+};
+&d1_dsp2 {
+	status = "okay";
+};
+
+&d0_dsp3 {
+	status = "okay";
+};
+&d1_dsp3 {
+	status = "okay";
+};
+
+&gpu0 {
+	status = "okay";
+};
+&d1_gpu {
+	status = "disabled";
+};
+
+&gc820 {
+	status = "okay";
+};
+&d1_gc820 {
+	status = "okay";
+};
+
+&vdec0 {
+	status = "okay";
+};
+&vdec1 {
+	status = "okay";
+};
+
+&venc0 {
+	status = "okay";
+};
+&venc1 {
+	status = "okay";
+};
+
+&video_output {
+	status = "okay";
+};
+
+&d1_video_output {
+	status = "okay";
+};
+
+&dc {
+	status = "okay";
+	dc_out: port {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		dc_out_dsi: endpoint@0 {
+			reg = <0>;
+			remote-endpoint = <&dsi_input>;
+		};
+		dc_out_dpi1: endpoint@1 {
+			reg = <1>;
+			remote-endpoint = <&vd_input>;
+		};
+		dc_out_hdmi: endpoint@2 {
+			reg = <2>;
+			remote-endpoint = <&hdmi_in_dc>;
+		};
+	};
+};
+
+&d1_dc {
+	status = "okay";
+	d1_dc_out: port {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		d1_dc_out_dsi: endpoint@0 {
+			reg = <0>;
+			remote-endpoint = <&d1_dsi_input>;
+		};
+		d1_dc_out_dpi1: endpoint@1 {
+			reg = <1>;
+			remote-endpoint = <&d1_vd_input>;
+		};
+		d1_dc_out_hdmi: endpoint@2 {
+			reg = <2>;
+			remote-endpoint = <&d1_hdmi_in_dc>;
+		};
+	};
+};
+
+&dc_test {
+	status = "disabled";
+};
+&d1_dc_test {
+	status = "disabled";
+};
+
+&virtual_display {
+	status = "okay";
+	port {
+		vd_input: endpoint {
+			remote-endpoint = <&dc_out_dpi1>;
+		};
+	};
+};
+
+&d1_virtual_display {
+	status = "okay";
+	port {
+		d1_vd_input: endpoint {
+			remote-endpoint = <&d1_dc_out_dpi1>;
+		};
+	};
+};
+
+&dsi_controller {
+    status = "okay";
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		port@0 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0>;
+			dsi_input: endpoint {
+				remote-endpoint = <&dc_out_dsi>;
+			};
+		};
+		port@1 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <1>;
+			mipi_dsi_out: endpoint {
+				remote-endpoint = <&panel_in>;
+			};
+		};
+	};
+
+	dsi_panel:dsi_panel@0 {
+		compatible = "eswin,generic-panel";
+		reg = <0>;
+		dsi,format = <0>; //RGB888:0, RGB666:1, RGB565:3
+		dsi,lanes = <4>;
+		port {
+			panel_in: endpoint {
+				remote-endpoint = <&mipi_dsi_out>;
+			};
+		};
+	};
+};
+&d1_dsi_controller {
+	status = "okay";
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		port@0 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0>;
+			d1_dsi_input: endpoint {
+				remote-endpoint = <&d1_dc_out_dsi>;
+			};
+		};
+		port@1 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <1>;
+			d1_mipi_dsi_out: endpoint {
+				remote-endpoint = <&d1_panel_in>;
+			};
+		};
+	};
+	d1_dsi_panel:dsi_panel@0 {
+		compatible = "eswin,generic-panel";
+		reg = <0>;
+		dsi,format = <0>; //RGB888:0, RGB666:1, RGB565:3
+		dsi,lanes = <4>;
+		port {
+			d1_panel_in: endpoint {
+				remote-endpoint = <&d1_mipi_dsi_out>;
+			};
+		};
+	};
+};
+
+&dsi_panel {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_die0_gpio27_default &pinctrl_die0_gpio106_default>;
+	backlight0-gpios = <&porta 27 GPIO_ACTIVE_HIGH>;
+	rst-gpios = <&portd 10 GPIO_ACTIVE_HIGH>;
+};
+&d1_dsi_panel {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_die1_gpio69_default &pinctrl_die1_gpio21_default>;
+	backlight0-gpios = <&d1_portc 5 GPIO_ACTIVE_HIGH>;
+	rst-gpios = <&d1_porta 21 GPIO_ACTIVE_HIGH>;
+};
+
+&dw_hdmi {
+	status = "okay";
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		port@0 {
+			reg = <0>;
+			hdmi_in_dc: endpoint@0 {
+				remote-endpoint = <&dc_out_hdmi>;
+			};
+		};
+		port@2 {
+			reg = <2>;
+			hdmi_in_i2s: endpoint@1 {
+				system-clock-frequency = <12288000>;
+				remote-endpoint = <&d0_i2s0_endpoint>;
+			};
+		};
+	};
+};
+&d1_dw_hdmi {
+	status = "okay";
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		port@0 {
+			reg = <0>;
+			d1_hdmi_in_dc: endpoint@0 {
+				remote-endpoint = <&d1_dc_out_hdmi>;
+			};
+		};
+		port@2 {
+			reg = <2>;
+			d1_hdmi_in_i2s: endpoint@1 {
+				system-clock-frequency = <12288000>;
+				remote-endpoint = <&d1_i2s0_endpoint>;
+			};
+		};
+	};
+};
+
+&dw_hdmi_hdcp2 {
+	status = "okay";
+};
+&d1_dw_hdmi_hdcp2 {
+	status = "disabled";
+};
+
+
+&d0_i2s0 {
+	status = "okay";
+	numa-node-id = <0>;
+	d0_i2s0_port: port {
+		d0_i2s0_endpoint: endpoint {
+			remote-endpoint = <&hdmi_in_i2s>;
+			dai-format = "i2s";
+		};
+	};
+};
+
+&d1_i2s0 {
+	status = "okay";
+	numa-node-id = <1>;
+	d1_i2s0_port: port {
+		d1_i2s0_endpoint: endpoint {
+			remote-endpoint = <&d1_hdmi_in_i2s>;
+			dai-format = "i2s";
+		};
+	};
+};
+
+&d0_i2s1 {
+	status = "okay";
+	numa-node-id = <0>;
+	d0_i2s1_port: port {
+		d0_i2s1_endpoint: endpoint {
+			remote-endpoint = <&d0_codec0_endpoint>;
+			dai-format = "i2s";
+		};
+	};
+};
+
+&d1_i2s1 {
+	status = "okay";
+	numa-node-id = <1>;
+	d1_i2s1_port: port {
+		d1_i2s1_endpoint: endpoint {
+			remote-endpoint = <&d1_codec0_endpoint>;
+			dai-format = "i2s";
+		};
+	};
+};
+
+&d0_i2s2 {
+	status = "okay";
+	numa-node-id = <0>;
+	d0_i2s2_port: port {
+		d0_i2s2_endpoint: endpoint {
+			remote-endpoint = <&d0_codec1_endpoint>;
+			dai-format = "i2s";
+		};
+	};
+};
+
+&d1_i2s2 {
+	status = "okay";
+	numa-node-id = <1>;
+	d1_i2s2_port: port {
+		d1_i2s2_endpoint: endpoint {
+			remote-endpoint = <&d1_codec1_endpoint>;
+			dai-format = "i2s";
+		};
+	};
+};
+
+&d0_graphcard0 {
+	status = "okay";
+	label = "Analog Audio-0";
+	dais = <&d0_i2s1_port>;
+};
+
+&d0_graphcard1 {
+	status = "okay";
+	label = "Analog Audio-1";
+	dais = <&d0_i2s2_port>;
+};
+
+&d0_graphcard2 {
+	status = "okay";
+	label = "HDMI Audio-0";
+	dais = <&d0_i2s0_port>;
+};
+
+&d1_graphcard0 {
+	status = "okay";
+	label = "Analog Audio-2";
+	dais = <&d1_i2s1_port>;
+};
+
+&d1_graphcard1 {
+	status = "okay";
+	label = "Analog Audio-3";
+	dais = <&d1_i2s2_port>;
+};
+
+&d1_graphcard2 {
+	status = "okay";
+	label = "HDMI Audio-1";
+	dais = <&d1_i2s0_port>;
+};
+
+&isp_0 {
+	status = "okay";
+};
+&d1_isp_0 {
+	status = "disabled";
+};
+
+&isp_1 {
+	status = "okay";
+};
+&d1_isp_1 {
+	status = "disabled";
+};
+
+&dewarp {
+	status = "okay";
+};
+&d1_dewarp {
+	status = "okay";
+};
+
+&mipi_dphy_rx {
+	status = "okay";
+};
+&d1_mipi_dphy_rx {
+	status = "disabled";
+};
+
+&csi_dma0 {
+	status = "okay";
+};
+&d1_csi_dma0 {
+	status = "disabled";
+};
+
+&csi_dma1 {
+	status = "disabled";
+};
+&d1_csi_dma1 {
+	status = "disabled";
+};
+
+&csi2_0 {
+	status = "okay";
+};
+&d1_csi2_0 {
+	status = "disabled";
+};
+
+&csi2_1 {
+	status = "disabled";
+};
+&d1_csi2_1 {
+	status = "disabled";
+};
+
+&sdhci_emmc {
+	/* emmc */
+	status = "okay";
+	delay_code = <0x17>;
+	drive-impedance-ohm = <50>;
+	enable-cmd-pullup;
+	enable-data-pullup;
+	no-sdio;
+	no-sd;
+};
+
+&d1_sdhci_emmc {
+	/* emmc */
+	status = "okay";
+	delay_code = <0x17>;
+	drive-impedance-ohm = <50>;
+	enable-cmd-pullup;
+	enable-data-pullup;
+	no-sdio;
+	no-sd;
+};
+
+&sdio0 {
+	/* sd card */
+	status = "okay";
+	delay_code = <0x55>;
+	drive-impedance-ohm = <33>;
+	enable-cmd-pullup;
+	enable-data-pullup;
+	no-sdio;
+	no-mmc;
+};
+
+&d1_sdio0 {
+	/* wifi module */
+	status = "okay";
+	delay_code = <0x29>;
+	drive-impedance-ohm = <33>;
+	enable-cmd-pullup;
+	enable-data-pullup;
+	keep-power-in-suspend;
+	non-removable;
+	no-sd;
+	no-mmc;
+	aw315:wifi_aw3155@0 {
+		compatible = "aml_w1_sdio";
+		reg = <0x0>;
+		interrupt-parent = <&d1_porta>;
+		interrupts = <18 IRQ_TYPE_LEVEL_LOW>;
+		pinctrl-names = "default", "default";
+		pinctrl-0 = <&pinctrl_die1_gpio18_default>;
+		pinctrl-1 = <&pinctrl_die1_gpio16_default>;
+		irq-gpios = <&d1_porta 18 GPIO_ACTIVE_HIGH>;
+		rst-gpios = <&d1_porta 16 GPIO_ACTIVE_HIGH>;
+	};
+};
+
+&sdio1 {
+	status = "disabled";
+};
+
+&d1_sdio1 {
+	status = "disabled";
+};
+
+&d0_gmac0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_die0_gpio15_default>;
+	rst-gpios = <&porta 15 GPIO_ACTIVE_LOW>;
+	eswin,rgmiisel = <&pinctrl 0x290 0x3>;
+	eswin,led-cfgs = <0x6100 0xa40 0x420>;
+	dly-param-1000m = <0x23232323 0x800c8023 0x0c0c0c0c>;
+	dly-param-100m = <0x50505050 0x803f8050 0x3f3f3f3f>;
+	dly-param-10m = <0 0 0>;
+	status = "okay";
+};
+
+&d1_gmac0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_die1_gpio13_default>;
+	rst-gpios = <&d1_porta 13 GPIO_ACTIVE_LOW>;
+	eswin,rgmiisel = <&d1_pinctrl 0x290 0x3>;
+	eswin,led-cfgs = <0x6100 0xa40 0x420>;
+	dly-param-1000m = <0x23232323 0x800c8023 0x0c0c0c0c>;
+	dly-param-100m = <0x50505050 0x803f8050 0x3f3f3f3f>;
+	dly-param-10m = <0 0 0>;
+	status = "okay";
+};
+
+&d0_gmac1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_die0_gpio14_default>;
+	rst-gpios = <&porta 14 GPIO_ACTIVE_LOW>;
+	eswin,rgmiisel = <&pinctrl 0x294 0x3>;
+	eswin,led-cfgs = <0x6100 0xa40 0x420>;
+	dly-param-1000m = <0x25252525 0x80268025 0x26262626>;
+	dly-param-100m = <0x48484848 0x80588048 0x58585858>;
+	dly-param-10m = <0 0 0>;
+	status = "okay";
+};
+
+&d1_gmac1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_die1_gpio14_default>;
+	rst-gpios = <&d1_porta 14 GPIO_ACTIVE_LOW>;
+	eswin,rgmiisel = <&d1_pinctrl 0x294 0x3>;
+	eswin,led-cfgs = <0x6100 0xa40 0x420>;
+	dly-param-1000m = <0x25252525 0x80268025 0x26262626>;
+	dly-param-100m = <0x48484848 0x80588048 0x58585858>;
+	dly-param-10m = <0 0 0>;
+	status = "okay";
+};
+
+&d0_sata {
+	status = "okay";
+};
+
+&d1_sata {
+	status = "okay";
+};
+
+&d0_usbdrd3_0 {
+	status = "okay";
+};
+&d1_usbdrd3_0 {
+	status = "okay";
+};
+
+&d0_usbdrd_dwc3_0 {
+	status = "okay";
+	dr_mode = "host";
+	maximum-speed = "super-speed";
+};
+
+&d1_usbdrd_dwc3_0 {
+	status = "okay";
+	dr_mode = "host";
+	maximum-speed = "super-speed";
+};
+
+&d0_usbdrd3_1 {
+	status = "okay";
+};
+
+&d1_usbdrd3_1 {
+	status = "okay";
+};
+
+&d0_usbdrd_dwc3_1 {
+	status = "okay";
+	dr_mode = "host";
+	maximum-speed = "super-speed";
+};
+
+&d1_usbdrd_dwc3_1 {
+	status = "okay";
+	dr_mode = "host";
+	maximum-speed = "super-speed";
+};
+
+&d0_dmac0 {
+	status = "okay";
+};
+
+&d1_dmac0 {
+	status = "disabled";
+};
+
+&d0_aon_dmac {
+	status = "okay";
+};
+
+&d1_aon_dmac {
+	status = "okay";
+};
+
+&d0_uart0 {
+	/* debug */
+	status = "okay";
+};
+
+&d1_uart0 {
+	status = "okay";
+};
+
+&d0_uart1 {
+	/* RS232 DB9 */
+	status = "okay";
+};
+
+&d1_uart1 {
+	/* BT M.2 KEY-E */
+	status = "okay";
+};
+
+&d0_uart2 {
+	/* pin header */
+	status = "okay";
+};
+
+&d1_uart2 {
+	/* pin header */
+	status = "okay";
+};
+
+&d0_uart3 {
+	/* unused */
+	status = "disabled";
+};
+
+&d1_uart3 {
+	/* unused */
+	status = "disabled";
+};
+
+&d0_uart4 {
+	/* unused */
+	status = "disabled";
+};
+
+&d1_uart4 {
+	/* unused */
+	status = "disabled";
+};
+
+&ssi0 {
+	/* spi flash */
+	status = "okay";
+	num-cs = <2>;
+	spi-flash@0 {
+		compatible = "winbond,w25q128fw",
+					"jedec,spi-nor";
+		reg = <0>;
+		#address-cells = <1>;
+		#size-cells = <1>;
+		spi-max-frequency = <4800000>;
+		rx-sample-delay-ns = <10>;
+	};
+	spi-flash@1 {
+		compatible = "winbond,w25q128fw",
+					"jedec,spi-nor";
+		reg = <1>;
+		#address-cells = <1>;
+		#size-cells = <1>;
+		spi-max-frequency = <4800000>;
+		rx-sample-delay-ns = <10>;
+	};
+};
+&d1_ssi0 {
+	/* spi flash */
+	status = "okay";
+	num-cs = <2>;
+	spi-flash@0 {
+		compatible = "winbond,w25q128fw",
+					"jedec,spi-nor";
+		reg = <0>;
+		#address-cells = <1>;
+		#size-cells = <1>;
+		spi-max-frequency = <4800000>;
+		rx-sample-delay-ns = <10>;
+	};
+	spi-flash@1 {
+		compatible = "winbond,w25q128fw",
+					"jedec,spi-nor";
+		reg = <1>;
+		#address-cells = <1>;
+		#size-cells = <1>;
+		spi-max-frequency = <4800000>;
+		rx-sample-delay-ns = <10>;
+	};
+};
+
+&ssi1 {
+	/* unused */
+	status = "disabled";
+};
+
+&d1_ssi1 {
+	/* unused */
+	status = "disabled";
+};
+
+&bootspi {
+	/* spi flash */
+	status = "okay";
+	num-cs = <1>;
+	cs-gpios = <&portd 0 GPIO_ACTIVE_LOW>;
+	wp-gpios = <&portd 4 GPIO_ACTIVE_LOW>;
+	spi-flash@0 {
+		compatible = "winbond,w25q128jw",
+					"jedec,spi-nor";
+		reg = <0>;
+		#address-cells = <1>;
+		#size-cells = <1>;
+		spi-max-frequency = <4800000>;
+		rx-sample-delay-ns = <10>;
+	};
+};
+
+&d1_bootspi {
+	/* spi flash */
+	status = "disabled";
+	num-cs = <1>;
+	cs-gpios = <&d1_portd 0 GPIO_ACTIVE_LOW>;
+	wp-gpios = <&d1_portd 4 GPIO_ACTIVE_LOW>;
+	spi-flash@0 {
+		compatible = "winbond,w25q128jw",
+					"jedec,spi-nor";
+		reg = <0>;
+		#address-cells = <1>;
+		#size-cells = <1>;
+		spi-max-frequency = <4800000>;
+		rx-sample-delay-ns = <10>;
+	};
+};
+
+&d0_mbox0 {
+	status = "okay";
+};
+&d1_mbox0 {
+	status = "okay";
+};
+
+&d0_mbox1 {
+	status = "okay";
+};
+&d1_mbox1 {
+	status = "okay";
+};
+
+&d0_mbox2 {
+	status = "okay";
+};
+&d1_mbox2 {
+	status = "okay";
+};
+
+&d0_mbox3 {
+	status = "okay";
+};
+&d1_mbox3 {
+	status = "okay";
+};
+
+&d0_mbox4 {
+	status = "okay";
+};
+&d1_mbox4 {
+	status = "okay";
+};
+
+&d0_mbox5 {
+	status = "okay";
+};
+&d1_mbox5 {
+	status = "okay";
+};
+
+&d0_mbox6 {
+	status = "okay";
+};
+&d1_mbox6 {
+	status = "okay";
+};
+
+&d0_mbox7 {
+	status = "okay";
+};
+&d1_mbox7 {
+	status = "okay";
+};
+
+&fan_control {
+	status = "okay";
+};
+&d1_fan_control {
+	status = "disabled";
+};
+
+&d0_i2c0 {
+	/* codec es8388 */
+	status = "okay";
+	d0_es8388_0: es8388-0@10 {
+		compatible = "eswin,es8388";
+		reg = <0x10>;
+		#sound-dai-cells = <0>;
+		port {
+			d0_codec0_endpoint: endpoint {
+				system-clock-frequency = <12288000>;
+				remote-endpoint = <&d0_i2s1_endpoint>;
+			};
+		};
+	};
+	d0_es8388_1: es8388-1@11 {
+		compatible = "eswin,es8388";
+		reg = <0x11>;
+		#sound-dai-cells = <0>;
+		port {
+			d0_codec1_endpoint: endpoint {
+				system-clock-frequency = <12288000>;
+				remote-endpoint = <&d0_i2s2_endpoint>;
+			};
+		};
+	};
+};
+
+&d1_i2c0 {
+	/* codec es8388 */
+	status = "okay";
+	d1_es8388_0: es8388-0@10 {
+		compatible = "eswin,es8388";
+		reg = <0x10>;
+		#sound-dai-cells = <0>;
+		port {
+			d1_codec0_endpoint: endpoint {
+				system-clock-frequency = <12288000>;
+				remote-endpoint = <&d1_i2s1_endpoint>;
+			};
+		};
+	};
+	d1_es8388_1: es8388-1@11 {
+		compatible = "eswin,es8388";
+		reg = <0x11>;
+		#sound-dai-cells = <0>;
+		port {
+			d1_codec1_endpoint: endpoint {
+				system-clock-frequency = <12288000>;
+				remote-endpoint = <&d1_i2s2_endpoint>;
+			};
+		};
+	};
+};
+
+&d0_i2c1 {
+	status = "okay";
+	es5430@f {
+		compatible = "einno,es5340";
+		reg = <0xf>;
+		eswin,regulator_default-microvolt=<1000000>;
+		eswin,regulator_label = "supply vdd1", "npu vdd1", "npu current1", "npu temperature1";
+		label = "npu_vdd";
+		regulators{
+			d0_npu_vcc:npu_svcc{
+				regulator-name="NPU_SVCC";
+				regulator-min-microvolt=<700000>;
+				regulator-max-microvolt=<1100000>;
+				regulator-min-microamp=<20000000>;
+				regulator-max-microamp=<40000000>;
+				regulator-ov-protection-microvolt=<1100000>;
+				regulator-always-on;
+			};
+		};
+	};
+};
+&d1_i2c1 {
+	status = "okay";
+	tps549d22@10 {
+		compatible = "ti,tps549d22";
+		reg = <0x10>;
+		eswin,regulator_default-microvolt=<1000000>;
+		label = "d1_npu_vdd";
+		regulators{
+			d1_npu_vcc:npu_svcc{
+				regulator-name="NPU_SVCC2";
+				regulator-min-microvolt=<700000>;
+				regulator-max-microvolt=<1100000>;
+				regulator-min-microamp=<20000000>;
+				regulator-max-microamp=<40000000>;
+				regulator-ov-protection-microvolt=<1100000>;
+			};
+		};
+	};
+};
+
+&d0_i2c2 {
+	/* touch screen*/
+	status = "ok";
+	d0_gt911:touchscreen@14 {
+		compatible = "goodix,gt911";
+		reg = <0x14>;
+		interrupt-parent = <&porta>;
+		interrupts = <20 IRQ_TYPE_EDGE_RISING>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_die0_gpio20_default>;
+		irq-gpios = <&porta 20 GPIO_ACTIVE_HIGH>;
+		reset-gpios = <&portd 15 GPIO_ACTIVE_HIGH>;
+	};
+};
+
+&d1_i2c2 {
+	/* touch screen*/
+	status = "ok";
+	d1_gt911:touchscreen@14 {
+		compatible = "goodix,gt911";
+		reg = <0x14>;
+		interrupt-parent = <&d1_porta>;
+		interrupts = <0 IRQ_TYPE_EDGE_RISING>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_die1_gpio0_default>;
+		irq-gpios = <&d1_porta 0 GPIO_ACTIVE_HIGH>;
+		reset-gpios = <&d1_portd 15 GPIO_ACTIVE_HIGH>;
+	};
+};
+
+&d0_i2c3 {
+	/* mipi csi0/csi1 */
+	status = "okay";
+};
+&d1_i2c3 {
+	status = "disabled";
+};
+
+&d0_i2c4 {
+	/* mipi csi2/csi3 */
+	status = "disabled";
+};
+&d1_i2c4 {
+	status = "disabled";
+};
+
+&d0_i2c5 {
+	/* mipi csi4/csi5 */
+	status = "disabled";
+};
+&d1_i2c5 {
+	status = "disabled";
+};
+
+&d0_i2c6 {
+	/* unused */
+	status = "disabled";
+};
+&d1_i2c6 {
+	status = "disabled";
+};
+
+&d0_i2c7 {
+	/* unused */
+	status = "disabled";
+};
+&d1_i2c7 {
+	status = "disabled";
+};
+
+&d0_i2c8 {
+	/* io extended for mipi csi */
+	status = "okay";
+	tca6416_0: gpio@20 {
+		compatible = "ti,tca6416";
+		reg = <0x20>;
+		gpio-controller; /* IRQ not connected */
+		#gpio-cells = <2>;
+		gpio-line-names = "MIPI_CSI0_PWDN", "MIPI_CSI0_RESET", "MIPI_CSI1_FBC", "MIPI_CSI1_ENB",
+				"MIPI_CSI1_RESET", "MIPI_CSI1_PWDN", "FREX_GP0", "",
+				"MIPI_CSI0_ENB", "MIPI_CSI0_FBC", "FREX_GP2", "MIPI_CSI2_FBC",
+				"MIPI_CSI2_ENB", "FREX_GP1", "MIPI_CSI2_RESET", "MIPI_CSI2_PWDN";
+	};
+	tca6416_1: gpio@21 {
+		compatible = "ti,tca6416";
+		reg = <0x21>;
+		gpio-controller;
+		#gpio-cells = <2>;
+		/* IRQ not connected */
+		gpio-line-names = "MIPI_CSI3_PWDN", "MIPI_CSI3_RESET", "MIPI_CSI3_ENB", "MIPI_CSI3_FBC",
+				"MIPI_CSI4_PWDN", "MIPI_CSI4_RESET", "MIPI_CSI4_ENB", "MIPI_CSI4_FBC",
+				"MIPI_CSI5_FBC", "MIPI_CSI5_ENB", "MIPI_CSI5_RESET", "MIPI_CSI5_PWDN",
+				"", "", "", "";
+	};
+};
+&d1_i2c8 {
+	status = "disabled";
+};
+
+&d0_i2c9 {
+	/* unused */
+	status = "disabled";
+};
+&d1_i2c9 {
+	status = "disabled";
+};
+
+&d0_aon_i2c0 {
+	status = "okay";
+	i2c-sda-hold-time-ns = <0x40>;
+	ina226@44 {
+			compatible = "ti,ina226";
+			#io-channel-cells = <1>;
+			label = "d0_vdd";
+			reg = <0x44>;
+			shunt-resistor = <1000>;
+		};
+	tmp102@48 {
+		compatible = "ti,tmp102";
+		reg = <0x48>;
+		label = "d0_board_tmp";
+		#thermal-sensor-cells = <1>;
+	};
+};
+
+
+&d1_aon_i2c0 {
+	/* ina226x4 */
+	status = "okay";
+	i2c-sda-hold-time-ns = <0x40>;
+	ina226@40 {
+			compatible = "ti,ina226";
+			#io-channel-cells = <1>;
+			label = "d1_vdd";
+			reg = <0x40>;
+			shunt-resistor = <1000>;
+		};
+	tmp102@48 {
+		compatible = "ti,tmp102";
+		reg = <0x48>;
+		label = "d1_board_tmp";
+		#thermal-sensor-cells = <1>;
+	};
+};
+&d1_aon_i2c1 {
+	status = "disabled";
+};
+
+&pwm0 {
+	/* fan */
+	status = "okay";
+};
+&d1_pwm0 {
+	status = "disabled";
+};
+
+&pvt0 {
+	status = "okay";
+};
+&d1_pvt0 {
+	status = "okay";
+};
+
+&pvt1 {
+	status = "okay";
+};
+&d1_pvt1 {
+	status = "okay";
+};
+
+&wdt0 {
+	status = "disabled";
+};
+&d1_wdt0 {
+	status = "disabled";
+};
+
+&wdt1 {
+	status = "disabled";
+};
+&d1_wdt1 {
+	status = "disabled";
+};
+
+&wdt2 {
+	status = "disabled";
+};
+&d1_wdt2 {
+	status = "disabled";
+};
+
+&wdt3 {
+	status = "disabled";
+};
+&d1_wdt3 {
+	status = "disabled";
+};
+
+&die0_rtc {
+	status = "okay";
+};
+&d1_rtc {
+	status = "okay";
+};
+
+&timer0 {
+	status = "okay";
+};
+&d1_timer0 {
+	status = "okay";
+};
+
+&timer1 {
+	status = "okay";
+};
+&d1_timer1 {
+	status = "okay";
+};
+
+&timer2 {
+	status = "okay";
+};
+&d1_timer2 {
+	status = "okay";
+};
+
+&timer3 {
+	status = "okay";
+};
+&d1_timer3 {
+	status = "okay";
+};
+
+&pinctrl {
+	status = "okay";
+};
+&d1_pinctrl {
+	status = "okay";
+};
+
+&gpio0 {
+	status = "okay";
+};
+&d1_gpio0 {
+	status = "okay";
+};
+
+&dev_llc_d0{
+	/* apply_npu_1G_freq; */
+	npu-supply=<&d0_npu_vcc>;
+	status = "okay";
+};
+&d1_llc_dev{
+	/* apply_npu_1G_freq; */
+	npu-supply=<&d1_npu_vcc>;
+	status = "okay";
+};
+
+&d2d {
+	status = "okay";
+};
+
+&d1_d2d {
+	status = "okay";
+};
+
+/* die0 GPIO Function Description
+
+	gpio0   : pcie prstn(I)
+	gpio11  : tmp alert(I)
+	gpio13  : tf card detect(I)
+	gpio14  : gphy1 resetn(O)
+	gpio15  : gphy0 resetn(O)
+	gpio18  : ina226 alert(I)
+	gpio20  : lcd touch int(I)
+	gpio21  : head phone plug/unplug detection2(I)
+	gpio27  : black light pwr_en(O)
+	gpio28  : head phone plug/unplug detection1(I)
+	gpio68  : fan pwm(O)
+	gpio69  : fan touch(I)
+	gpio92  : frex gp0(O)
+	gpio93  : frex gp1(O)
+	gpio106 : mipi dsi resetn(O)
+	gpio111 : lcd touch resetn(O)
+*/
+/* die1 GPIO Function Description
+
+	gpio0   : lcd touch int(I)
+	gpio13  : gphy0 resetn(O)
+	gpio14  : gphy1 resetn(O)
+	gpio15  : bt wake host(I)
+	gpio16  : sdio wifi disable(O)
+	gpio18  : sdio wifi wake host(I)
+	gpio19  : pcie prstn(I)
+	gpio20  : head phone plug/unplug detection2(I)
+	gpio21  : mipi dsi resetn(O)
+	gpio27  : tmp alert(I)
+	gpio28  : head phone plug/unplug detection1(I)
+	gpio68  : system led(O)
+	gpio69  : black light pwr_en(O)
+	gpio92  : frex gp0(O)
+	gpio93  : frex gp1(O)
+	gpio106 : host wak bt(O)
+	gpio111 : lcd touch resetn(O)
+*/
diff --git a/arch/riscv/configs/eic7700_defconfig b/arch/riscv/configs/eic7700_defconfig
index f621172704ea..72cc0c1c597c 100644
--- a/arch/riscv/configs/eic7700_defconfig
+++ b/arch/riscv/configs/eic7700_defconfig
@@ -466,6 +466,7 @@ CONFIG_GPIO_DWAPB=y
 CONFIG_SENSORS_ESWIN_FAN_CONTROL=y
 CONFIG_SENSORS_ESWIN_PVT=y
 CONFIG_SENSORS_INA2XX=y
+CONFIG_SENSORS_TMP102=y
 CONFIG_SENSORS_PAC1934=y
 CONFIG_THERMAL=y
 CONFIG_THERMAL_STATISTICS=y
@@ -478,8 +479,9 @@ CONFIG_THERMAL_EMULATION=y
 CONFIG_WATCHDOG=y
 CONFIG_DW_WATCHDOG=y
 CONFIG_REGULATOR=y
-CONFIG_REGULATOR_ES5340=y
 CONFIG_REGULATOR_MPQ8785=y
+CONFIG_REGULATOR_TPS549D22=y
+CONFIG_REGULATOR_ES5340=y
 CONFIG_REGULATOR_PCA9450=y
 CONFIG_MEDIA_SUPPORT=y
 CONFIG_MEDIA_SUPPORT_FILTER=y
diff --git a/arch/riscv/configs/eic7702_defconfig b/arch/riscv/configs/eic7702_defconfig
index 7b5fa46fad2a..d86411ce75d4 100644
--- a/arch/riscv/configs/eic7702_defconfig
+++ b/arch/riscv/configs/eic7702_defconfig
@@ -467,6 +467,7 @@ CONFIG_WATCHDOG=y
 CONFIG_DW_WATCHDOG=y
 CONFIG_REGULATOR=y
 CONFIG_REGULATOR_ES5340=y
+CONFIG_REGULATOR_TPS549D22=y
 CONFIG_REGULATOR_PCA9450=y
 CONFIG_MEDIA_SUPPORT=y
 CONFIG_MEDIA_SUPPORT_FILTER=y
diff --git a/drivers/regulator/Kconfig b/drivers/regulator/Kconfig
index 1c469ddbe464..661407848b3c 100644
--- a/drivers/regulator/Kconfig
+++ b/drivers/regulator/Kconfig
@@ -799,7 +799,12 @@ config REGULATOR_MPQ8785
 	tristate "mps,mpq8785"
 	help
 	  This driver provides support for the voltage regulators on the
-	  eswin evb.
+	  eswin board.
+config REGULATOR_TPS549D22
+	tristate "ti,tps549d22"
+	help
+	  This driver provides support for the voltage regulators on the
+	  eswin board.
 config REGULATOR_ES5340
 	tristate "einno,es5340"
 	depends on I2C && OF
diff --git a/drivers/regulator/Makefile b/drivers/regulator/Makefile
index c182cbff6548..cdc632121ada 100644
--- a/drivers/regulator/Makefile
+++ b/drivers/regulator/Makefile
@@ -197,4 +197,5 @@ obj-$(CONFIG_REGULATOR_WM8400) += wm8400-regulator.o
 obj-$(CONFIG_REGULATOR_WM8994) += wm8994-regulator.o
 obj-$(CONFIG_REGULATOR_MPQ8785) += mpq8785.o
 obj-$(CONFIG_REGULATOR_ES5340) += es5340.o
+obj-$(CONFIG_REGULATOR_TPS549D22) += tps549d22-regulator.o
 ccflags-$(CONFIG_REGULATOR_DEBUG) += -DDEBUG
diff --git a/drivers/regulator/tps549d22-regulator.c b/drivers/regulator/tps549d22-regulator.c
new file mode 100644
index 000000000000..1f247c14a50b
--- /dev/null
+++ b/drivers/regulator/tps549d22-regulator.c
@@ -0,0 +1,580 @@
+
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Regulator driver for TI TPS549D22
+ *
+* Copyright 2024, Beijing ESWIN Computing Technology Co., Ltd.. All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Authors: Yang Wei <yangwei1@eswincomputing.com>
+ */
+
+#include <linux/cdev.h>
+#include <linux/device.h>
+#include <linux/err.h>
+#include <linux/fs.h>
+#include <linux/gpio/consumer.h>
+#include <linux/hwmon-sysfs.h>
+#include <linux/hwmon.h>
+#include <linux/i2c.h>
+#include <linux/init.h>
+#include <linux/interrupt.h>
+#include <linux/irq.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/regulator/driver.h>
+#include <linux/regulator/machine.h>
+#include <linux/regulator/of_regulator.h>
+#include <linux/slab.h>
+#include <linux/sysfs.h>
+
+#define TPS549D22_CMD_OPERATION 0x1
+#define TPS549D22_CMD_ON_OFF_CONFIG 0x2
+#define TPS549D22_CMD_CLEAR_FAULT 0x3
+#define TPS549D22_CMD_WRITE_PROTECTION 0x10
+#define TPS549D22_CMD_CAPABILITY 0x19
+#define TPS549D22_CMD_PMBUS_PS_NUM 0x1C
+#define TPS549D22_CMD_VOUT_MODE 0x20
+#define TPS549D22_CMD_VOUT_COMMAND 0x21
+#define TPS549D22_CMD_VOUT_MARGIN_HIGH 0x25
+#define TPS549D22_CMD_VOUT_MARGIN_LOW 0x26
+#define TPS549D22_CMD_STATUS_BYTE 0x78
+#define TPS549D22_CMD_STATUS_WORD 0x79
+#define TPS549D22_CMD_STATUS_VOUT 0x7A
+#define TPS549D22_CMD_STATUS_IOUT 0x7B
+#define TPS549D22_CMD_STATUS_CML 0x7E
+
+#define TPS549D22_VOLT_STEP 2170 // uV
+#define TPS549D22_MASK_VOUT_VALUE 0x3FF
+#define TPS549D22_MASK_OPERATION_ENABLE 0X80
+
+struct TPS549D22_DRIVER_DATA
+{
+	struct regulator_dev *rdev;
+	struct i2c_client *client;
+	struct mutex config_lock;
+};
+
+static struct of_regulator_match tps549d22_matches[] = {
+	{
+		.name = "npu_svcc",
+	},
+};
+
+static inline u32 tps549d22_volt2reg(u32 vlot_uv)
+{
+	u32 value = 0;
+
+	value = DIV_ROUND_CLOSEST(vlot_uv, TPS549D22_VOLT_STEP);
+	return value;
+}
+
+static inline s32 tps549d22_str2ul(const char *buf, u32 *value)
+{
+	unsigned long cache = 0;
+	int ret = 0;
+
+	if (NULL == strstr(buf, "0x"))
+	{
+		ret = kstrtoul(buf, 10, &cache);
+	}
+	else
+	{
+		ret = kstrtoul(buf, 16, &cache);
+	}
+	*value = cache;
+
+	return ret;
+}
+
+static u8 tps549d22_read_byte(struct TPS549D22_DRIVER_DATA *data, u8 command)
+{
+	int ret = 0;
+	mutex_lock(&data->config_lock);
+	ret = i2c_smbus_read_byte_data(data->client, command);
+	mutex_unlock(&data->config_lock);
+	if (ret < 0)
+	{
+		dev_err(&data->client->dev, "get command:0x%x value error:%d\n", command,
+				ret);
+		return 0xff;
+	}
+	return (u8)ret;
+}
+
+static s32 tps549d22_write_byte(struct TPS549D22_DRIVER_DATA *data, u8 command,
+								u8 val)
+{
+	int ret = 0;
+	mutex_lock(&data->config_lock);
+	ret = i2c_smbus_write_byte_data(data->client, command, val);
+	mutex_unlock(&data->config_lock);
+	if (ret < 0)
+	{
+		dev_err(&data->client->dev, "set command:0x%x value:0x%x error:%d\n",
+				command, val, ret);
+	}
+	return ret;
+}
+
+static s32 tps549d22_update_byte(struct TPS549D22_DRIVER_DATA *data, u8 command,
+								 u8 mask, u8 val)
+{
+	u8 old_value = 0;
+	u8 new_value = 0;
+	if (0 != (~mask & val))
+	{
+		dev_err(&data->client->dev, "command:0x%x,input:0x%x outrange mask:0x%x\n",
+				command, val, mask);
+		return -EINVAL;
+	}
+	old_value = tps549d22_read_byte(data, command);
+	new_value = ~mask & old_value;
+	new_value = new_value | val;
+	return tps549d22_write_byte(data, command, new_value);
+}
+
+static u16 tps549d22_read_word(struct TPS549D22_DRIVER_DATA *data, u8 command)
+{
+	int ret = 0;
+	mutex_lock(&data->config_lock);
+	ret = i2c_smbus_read_word_data(data->client, command);
+	mutex_unlock(&data->config_lock);
+	if (ret < 0)
+	{
+		dev_err(&data->client->dev, "get command:0x%x value error:%d\n", command,
+				ret);
+		return 0xffff;
+	}
+	return (u16)ret;
+}
+
+static u16 tps549d22_read_mask_word(struct TPS549D22_DRIVER_DATA *data, u8 command,
+									u16 mask)
+{
+	u16 ret = tps549d22_read_word(data, command);
+	return (ret & mask);
+}
+
+static s32 tps549d22_write_word(struct TPS549D22_DRIVER_DATA *data, u8 command,
+								u16 val)
+{
+	int ret = 0;
+	mutex_lock(&data->config_lock);
+	ret = i2c_smbus_write_word_data(data->client, command, val);
+	mutex_unlock(&data->config_lock);
+	if (ret < 0)
+	{
+		dev_err(&data->client->dev, "set command:0x%x value:0x%x error:%d\n",
+				command, val, ret);
+	}
+	return ret;
+}
+
+static s32 tps549d22_update_word(struct TPS549D22_DRIVER_DATA *data, u8 command,
+								 u16 mask, u16 val)
+{
+	u16 old_value = 0;
+	u16 new_value = 0;
+	if (0 != (~mask & val))
+	{
+		dev_err(&data->client->dev, "command:0x%x,input:0x%x outrange mask:0x%x\n",
+				command, val, mask);
+		return -EINVAL;
+	}
+	old_value = tps549d22_read_word(data, command);
+	new_value = ~mask & old_value;
+	new_value = new_value | val;
+	return tps549d22_write_word(data, command, new_value);
+}
+
+static int tps549d22_get_enable(struct TPS549D22_DRIVER_DATA *data)
+{
+	u8 cache = 0;
+
+	cache = tps549d22_read_byte(data, TPS549D22_CMD_OPERATION);
+
+	return ((cache >> 7) & 0x1);
+}
+
+static u32 tps549d22_get_vout(struct TPS549D22_DRIVER_DATA *data)
+{
+	u32 get_value = tps549d22_read_mask_word(data, TPS549D22_CMD_VOUT_COMMAND,
+											 TPS549D22_MASK_VOUT_VALUE);
+	return (get_value * TPS549D22_VOLT_STEP / 1000);
+}
+
+static s32 tps549d22_set_vout(struct TPS549D22_DRIVER_DATA *data, u32 volt_uv)
+{
+	u16 new_value = tps549d22_volt2reg(volt_uv);
+
+	const struct regulation_constraints *constraints = &tps549d22_matches[0].init_data->constraints;
+
+	if ((volt_uv > (constraints->max_uV)) || (volt_uv < (constraints->min_uV)))
+	{
+		dev_err(&data->rdev->dev, "max:%duV,min:%duV,now:%duV\n",
+				(constraints->max_uV), constraints->min_uV, volt_uv);
+		return -EINVAL;
+	}
+
+	return tps549d22_update_word(data, TPS549D22_CMD_VOUT_COMMAND,
+								 TPS549D22_MASK_VOUT_VALUE, (new_value));
+}
+
+int tps549d22_regulator_enable(struct regulator_dev *rdev)
+{
+	struct i2c_client *client = to_i2c_client(rdev->dev.parent);
+	struct TPS549D22_DRIVER_DATA *data = i2c_get_clientdata(client);
+	dev_dbg(&rdev->dev, "%s.%d\n", __FUNCTION__, __LINE__);
+	return tps549d22_update_byte(data, TPS549D22_CMD_OPERATION,
+								 TPS549D22_MASK_OPERATION_ENABLE,
+								 TPS549D22_MASK_OPERATION_ENABLE);
+}
+
+int tps549d22_regulator_disable(struct regulator_dev *rdev)
+{
+	struct i2c_client *client = to_i2c_client(rdev->dev.parent);
+	struct TPS549D22_DRIVER_DATA *data = i2c_get_clientdata(client);
+	dev_dbg(&rdev->dev, "%s.%d\n", __FUNCTION__, __LINE__);
+	return tps549d22_update_byte(data, TPS549D22_CMD_OPERATION,
+								 TPS549D22_MASK_OPERATION_ENABLE, 0);
+}
+int tps549d22_regulator_is_enabled(struct regulator_dev *rdev)
+{
+	struct i2c_client *client = to_i2c_client(rdev->dev.parent);
+	struct TPS549D22_DRIVER_DATA *data = i2c_get_clientdata(client);
+	dev_dbg(&rdev->dev, "%s.%d\n", __FUNCTION__, __LINE__);
+	return tps549d22_get_enable(data);
+}
+
+static struct linear_range tps549d22_ext_ranges[] = {
+	REGULATOR_LINEAR_RANGE(666000, 307, 614, TPS549D22_VOLT_STEP),
+};
+
+/**
+ * tps549d22_set_voltage_sel -  set_voltage_sel for users
+ *
+ * @rdev: regulator to operate on
+ * @sel: Selector to set
+ */
+static s32 tps549d22_set_voltage_sel(struct regulator_dev *rdev,
+									 unsigned selector)
+{
+	struct device *dev = &rdev->dev;
+	struct i2c_client *client = to_i2c_client(rdev->dev.parent);
+	struct TPS549D22_DRIVER_DATA *data = i2c_get_clientdata(client);
+	u32 new_value = 0;
+
+	if (selector > tps549d22_ext_ranges->max_sel)
+	{
+		dev_err(dev, "selector:%u out of rang 0~%u\n", selector,
+				tps549d22_ext_ranges->max_sel);
+		return -EINVAL;
+	}
+
+	new_value = tps549d22_ext_ranges->step * selector + tps549d22_ext_ranges->min;
+
+	dev_dbg(dev, "%s_volt:%duV,selector:%u,step:%u,min:%u\n", __FUNCTION__,
+			new_value, selector, tps549d22_ext_ranges->step,
+			tps549d22_ext_ranges->min);
+
+	tps549d22_set_vout(data, new_value);
+
+	return 0;
+}
+
+/**
+ * tps549d22_get_voltage_sel -  get_voltage_sel for users
+ *
+ * @rdev: regulator to operate on
+ */
+static s32 tps549d22_get_voltage_sel(struct regulator_dev *rdev)
+{
+	s32 index = 0;
+	struct device *dev = &rdev->dev;
+	struct i2c_client *client = to_i2c_client(rdev->dev.parent);
+	struct TPS549D22_DRIVER_DATA *data = i2c_get_clientdata(client);
+
+	index = tps549d22_read_mask_word(data, TPS549D22_CMD_VOUT_COMMAND, TPS549D22_MASK_VOUT_VALUE);
+	index = index - DIV_ROUND_CLOSEST(tps549d22_ext_ranges->min, TPS549D22_VOLT_STEP);
+	dev_dbg(dev, "%s index:%d\n", __FUNCTION__, index);
+	return index;
+}
+
+static s32 tps549d22_get_error_flags(struct regulator_dev *rdev, u32 *flags)
+{
+	struct i2c_client *client = to_i2c_client(rdev->dev.parent);
+	struct TPS549D22_DRIVER_DATA *data = i2c_get_clientdata(client);
+
+	*flags = tps549d22_read_byte(data, TPS549D22_CMD_STATUS_WORD);
+
+	dev_dbg(&rdev->dev, "tps549d22_get_error_flags_%u\n", *flags);
+	return 0;
+}
+
+
+
+int tps549d22_list_voltage_linear_range(struct regulator_dev *rdev,
+					unsigned int selector)
+{
+	int value = 0;
+	value = regulator_desc_list_voltage_linear_range(rdev->desc, selector);
+	value = (value/2000)*2000;
+	return value;
+}
+
+static struct regulator_ops tps549d22_core_ops = {
+
+	.list_voltage = tps549d22_list_voltage_linear_range,
+	.map_voltage = regulator_map_voltage_linear_range,
+	.set_voltage_sel = tps549d22_set_voltage_sel,
+	.get_voltage_sel = tps549d22_get_voltage_sel,
+	/* enable/disable regulator */
+	.enable = tps549d22_regulator_enable,
+	.disable = tps549d22_regulator_disable,
+	.is_enabled = tps549d22_regulator_is_enabled,
+	.get_error_flags = tps549d22_get_error_flags,
+
+};
+static struct regulator_desc tps549d22_regulator_desc = {
+	.name = "NPUVDD",
+	.type = REGULATOR_VOLTAGE,
+	.n_voltages = 307,
+	.ops = &tps549d22_core_ops,
+	.linear_ranges = tps549d22_ext_ranges,
+	.n_linear_ranges = ARRAY_SIZE(tps549d22_ext_ranges),
+	.owner = THIS_MODULE,
+};
+
+static ssize_t tps549d22_vout_show(struct device *d,
+								   struct device_attribute *attr, char *buf)
+{
+	struct i2c_client *client = to_i2c_client(d);
+	struct TPS549D22_DRIVER_DATA *data = i2c_get_clientdata(client);
+
+	return sysfs_emit(buf, "%u", tps549d22_get_vout(data));
+}
+static ssize_t tps549d22_vout_store(struct device *d,
+									struct device_attribute *attr,
+									const char *buf, size_t count)
+{
+	struct i2c_client *client = to_i2c_client(d);
+	struct TPS549D22_DRIVER_DATA *data = i2c_get_clientdata(client);
+	u32 volt_value = 0;
+	int ret = 0;
+	ret = tps549d22_str2ul(buf, &volt_value);
+
+	if (ret)
+	{
+		return ret;
+	}
+	ret = tps549d22_set_vout(data, volt_value * 1000);
+	if (0 != ret)
+	{
+		return ret;
+	}
+	return count;
+}
+DEVICE_ATTR(tps549d22_vout, 0600, tps549d22_vout_show, tps549d22_vout_store);
+
+static ssize_t tps549d22_enable_show(struct device *d,
+									 struct device_attribute *attr, char *buf)
+{
+	struct i2c_client *client = to_i2c_client(d);
+	struct TPS549D22_DRIVER_DATA *data = i2c_get_clientdata(client);
+
+	return sysfs_emit(buf, "%u", data->rdev->desc->ops->is_enabled(data->rdev));
+}
+static ssize_t tps549d22_enable_store(struct device *d,
+									  struct device_attribute *attr,
+									  const char *buf, size_t count)
+{
+	struct i2c_client *client = to_i2c_client(d);
+	struct TPS549D22_DRIVER_DATA *data = i2c_get_clientdata(client);
+	u32 value = 0;
+	int ret = 0;
+	ret = tps549d22_str2ul(buf, &value);
+
+	if (ret)
+	{
+		return ret;
+	}
+	if (0 == value)
+	{
+		data->rdev->desc->ops->disable(data->rdev);
+	}
+	else
+	{
+		data->rdev->desc->ops->enable(data->rdev);
+	}
+	return count;
+}
+DEVICE_ATTR(tps549d22_enable, 0600, tps549d22_enable_show, tps549d22_enable_store);
+
+static ssize_t tps549d22_status_show(struct device *d,
+									 struct device_attribute *attr, char *buf)
+{
+	struct i2c_client *client = to_i2c_client(d);
+	struct TPS549D22_DRIVER_DATA *data = i2c_get_clientdata(client);
+
+	return sysfs_emit(buf, "STATUS_BYTE:0x%x\nSTATUS_WORD:0x%x\n"
+						   "STATUS_VOUT:0x%x\nSTATUS_IOUT:0x%x\nSTATUS_CML:0x%x\n",
+					  tps549d22_read_byte(data, TPS549D22_CMD_STATUS_BYTE),
+					  tps549d22_read_byte(data, TPS549D22_CMD_STATUS_WORD),
+					  tps549d22_read_byte(data, TPS549D22_CMD_STATUS_VOUT),
+					  tps549d22_read_byte(data, TPS549D22_CMD_STATUS_IOUT),
+					  tps549d22_read_byte(data, TPS549D22_CMD_STATUS_CML));
+}
+DEVICE_ATTR_RO(tps549d22_status);
+
+static struct attribute *tps549d22_attrs[] = {
+	&dev_attr_tps549d22_vout.attr, &dev_attr_tps549d22_enable.attr,
+	&dev_attr_tps549d22_status.attr, NULL};
+
+ATTRIBUTE_GROUPS(tps549d22);
+
+static s32 tps549d22_init_data(struct TPS549D22_DRIVER_DATA *data,
+							   const struct regulation_constraints *constraints, u32 default_voltage)
+{
+	s32 ret = 0;
+	struct device *dev = &data->client->dev;
+
+	tps549d22_ext_ranges->min = constraints->min_uV;
+	tps549d22_ext_ranges->min_sel = 0;
+	tps549d22_ext_ranges->max_sel = DIV_ROUND_CLOSEST(constraints->max_uV - constraints->min_uV, TPS549D22_VOLT_STEP) + 1;
+	tps549d22_regulator_desc.n_voltages = tps549d22_ext_ranges->max_sel - tps549d22_ext_ranges->min_sel;
+	dev_dbg(dev,
+			"min_uV:%d,max_uV:%d,min_sel:%d,max_sel:%d,n_voltages:%d\n",
+			tps549d22_ext_ranges->min, constraints->max_uV,
+			tps549d22_ext_ranges->min_sel, tps549d22_ext_ranges->max_sel,
+			tps549d22_regulator_desc.n_voltages);
+
+	tps549d22_set_vout(data, default_voltage);
+	tps549d22_write_byte(data, TPS549D22_CMD_ON_OFF_CONFIG, 0xf);
+	return ret;
+}
+
+static s32 tps549d22_probe(struct i2c_client *client)
+{
+	struct TPS549D22_DRIVER_DATA *data = NULL;
+	s32 ret = 0;
+	s32 regulator_cnt = 0;
+	u32 default_voltage = 0;
+	struct device *hwmon_dev;
+	struct regulator_config config = {};
+	struct device *dev = &client->dev;
+	struct device_node *np, *parent;
+
+	if (!i2c_check_functionality(client->adapter, I2C_FUNC_SMBUS_BYTE_DATA))
+	{
+		dev_err(dev, "not support smbus\n");
+		return -EIO;
+	}
+	data = devm_kzalloc(dev, sizeof(struct TPS549D22_DRIVER_DATA), GFP_KERNEL);
+	if (!data)
+		return -ENOMEM;
+
+	mutex_init(&data->config_lock);
+	data->client = client;
+	i2c_set_clientdata(client, data);
+	/* Get the device (PMIC) node */
+	np = of_node_get(dev->of_node);
+	if (!np)
+		return -EINVAL;
+
+	/* Get 'regulators' subnode */
+	parent = of_get_child_by_name(np, "regulators");
+	if (!parent)
+	{
+		dev_err(dev, "regulators node not found\n");
+		return -EINVAL;
+	}
+	ret = of_property_read_u32(np, "eswin,regulator_default-microvolt", &default_voltage);
+	if (ret)
+	{
+		default_voltage = 900000;
+	}
+
+	/* fill isl6271a_matches array */
+	regulator_cnt = of_regulator_match(dev, parent, tps549d22_matches, ARRAY_SIZE(tps549d22_matches));
+	of_node_put(parent);
+	if (regulator_cnt != 1)
+	{
+		dev_err(dev, "Error parsing regulator init data: %d\n", regulator_cnt);
+		return regulator_cnt;
+	}
+
+	/* Fetched from device tree */
+	config.init_data = tps549d22_matches[0].init_data;
+	config.dev = dev;
+	config.of_node = tps549d22_matches[0].of_node;
+	/* config.ena_gpio = -EINVAL; */
+	ret = tps549d22_init_data(data, &config.init_data->constraints, default_voltage);
+	if (0 != ret)
+	{
+		dev_err(dev, "init tps549d22 error\n");
+		return -EIO;
+	}
+	data->rdev = devm_regulator_register(dev, &tps549d22_regulator_desc, &config);
+	if (IS_ERR(data->rdev))
+	{
+		dev_err(dev, "failed to register %s\n", tps549d22_regulator_desc.name);
+	}
+	hwmon_dev = devm_hwmon_device_register_with_groups(
+		dev, client->name, data, tps549d22_groups);
+	if (IS_ERR(hwmon_dev))
+		return PTR_ERR(hwmon_dev);
+
+	dev_info(dev, "tps549d22_probe\n");
+
+	return 0;
+}
+
+static void tps549d22_remove(struct i2c_client *client)
+{
+	dev_info(&client->dev, "tps549d22_remove\n");
+}
+
+static s32 tps549d22_detect(struct i2c_client *client,
+							struct i2c_board_info *info)
+{
+	dev_info(&client->dev, "tps549d22_detect\n");
+	return 0;
+}
+
+static const struct i2c_device_id tps549d22_id[] = {{"tps549d22", 0}, {}};
+MODULE_DEVICE_TABLE(i2c, tps549d22_id);
+
+/* Addresses to scan */
+static const unsigned short normal_i2c[] = {0x2c, 0x2d, 0x2e, 0x60,
+											I2C_CLIENT_END};
+
+static struct i2c_driver tps549d22_driver = {
+	.class = I2C_CLASS_HWMON,
+	.driver =
+		{
+			.name = "tps549d22",
+		},
+	.probe = tps549d22_probe,
+	.remove = tps549d22_remove,
+	.id_table = tps549d22_id,
+	.detect = tps549d22_detect,
+	.address_list = normal_i2c,
+};
+
+module_i2c_driver(tps549d22_driver);
+
+MODULE_AUTHOR("Yang Wei <yangwei1@eswincomputing.com>");
+MODULE_DESCRIPTION("tps549d22 driver");
+MODULE_LICENSE("GPL");
-- 
2.48.1

